File: 0542-01-matrix/0542-01-matrix.java
class Solution {
    public int[][] updateMatrix(int[][] mat) {
        // use a matrix to store the result
        // if mat[i][j]==0 res[i][j]=0;
        // else use BFS to find the shortest distance of it to 0;
        // 1st pass
        int m = mat.length;
        int n = mat[0].length;
        
        // use BFS to solve it
        Queue<int[]> queue = new LinkedList<>();
        
        // Initialize the result matrix with max values for 1s, and add them into queue
        int[][] result = new int[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (mat[i][j] == 0) {
                    result[i][j] = 0;
                    queue.offer(new int[]{i,j}); // store coordinate into it
                } else {
                    result[i][j] = Integer.MAX_VALUE; // Large value for 1s
                }
            }
        }
        // helper directions
        int[][] directions = {{1, 0},{-1, 0},{0, 1},{0, -1}};
        
        // BFS
        while(!queue.isEmpty()){
            // get an int[]
            int[] cell = queue.poll();
            int x = cell[0];
            int y = cell[1];
            
            // and compare it with 4 directions
            for(int[] direction: directions){
                int newX = x + direction[0];
                int newY = y + direction[1];
             // if 4 directions > current value + 1, update 4 directions value, add it to queue 
                if(newX>=0 && newX<mat.length && newY >=0 && newY<mat[0].length){
                    if(result[newX][newY] > result[x][y] + 1){
                        result[newX][newY] = result[x][y] + 1;
                        queue.offer(new int[]{newX, newY});
                    }
                }
            }
        }

        return result;
    }
}

File: 0542-01-matrix/0542-01-matrix.py
class Solution:
    def updateMatrix(self, mat: List[List[int]]) -> List[List[int]]:
        # BFS
        m = len(mat)
        n = len(mat[0])
        result = [[float('inf')] * n for _ in range(m)]  # Initialize result matrix with infinity
        
        queue = deque()
        
        # initialize result and queue
        for i in range (m):
            for j in range (n):
                if mat[i][j]==0:
                    result[i][j]=0
                    queue.append((i,j))
        
        # helper list
        directions = [(1, 0), (-1, 0), (0, 1), (0, -1)]
        # BFS
        while queue:
            # get en element
            x, y = queue.popleft() 
            # update 4 directions
            for dx, dy in directions:
                newX, newY = x + dx, y + dy
                # check validation
                if 0 <= newX < m and 0 <= newY < n:
                    if result[newX][newY] > result[x][y] + 1:
                        result[newX][newY] = result[x][y] + 1
                        queue.append((newX, newY))
                        
        return result
        